runtime/reconcilers/model.go

```diff
@@ -603,118 +612,631 @@ func (r *ModelReconciler) resolveIncrementalState(ctx context.Context, mdl *runt
+func (r *ModelReconciler) syncSplits(ctx context.Context, mdl *runtimev1.ModelV2, startIdx int, rows []map[string]any) error {
+	if len(rows) == 0 {
+		return nil
+	}
+
+	catalog, release, err := r.C.Runtime.Catalog(ctx, r.C.InstanceID)
+	if err != nil {
+		return err
 	}
 	defer release()
 
-	res, err := e.Execute(ctx)
+	// Build ModelSplit objects indexed by their Key
+	splits := make(map[string]drivers.ModelSplit, len(rows))
+	for i, row := range rows {
+		// If a watermark field is configured, we extract and remove it from the map.
+		// It is necessary to remove it to ensure the key is deterministic.
+		var watermark *time.Time
+		if mdl.Spec.SplitsWatermarkField != "" {
+			if v, ok := row[mdl.Spec.SplitsWatermarkField]; ok {
+				t, ok := v.(time.Time)
+				if !ok {
+					return fmt.Errorf(`expected a timestamp for split watermark field %q, got type %T`, mdl.Spec.SplitsWatermarkField, v)
+				}
+
+				watermark = &t
+				delete(row, mdl.Spec.SplitsWatermarkField)
+			}
+		}
+
+		// Marshal the rest of the row
+		rowJSON, err := json.Marshal(row)
+		if err != nil {
+			return fmt.Errorf("failed to marshal split row at index %d: %w", i, err)
+		}
+
+		// JSON serialization is deterministic in Go, so we can hash it to get a key
+		key, err := md5Hash(rowJSON)
+		if err != nil {
+			return fmt.Errorf("failed to hash split row at index %d: %w", i, err)
+		}
+
+		splits[key] = drivers.ModelSplit{
+			Key:        key,
+			DataJSON:   rowJSON,
+			Index:      startIdx + i,
+			Watermark:  watermark,
+			ExecutedOn: nil,
+			Error:      "",
+			Elapsed:    0,
+		}
+	}
+
+	// Find those splits that already exist in the catalog
+	keys := make([]string, 0, len(splits))
+	for key := range splits {
+		keys = append(keys, key)
+	}
+	existing, err := catalog.FindModelSplitsByKeys(ctx, mdl.State.SplitsModelId, keys)
 	if err != nil {
-		return "", nil, &modelBuildError{err: err}
+		return fmt.Errorf("failed to find existing splits: %w", err)
 	}
-	return executorName, res, nil
+
+	// Handle the existing splits by skipping or updating them.
+	// We remove the handled splits from the splits map. The ones that remain are new and should be inserted.
+	for _, old := range existing {
+		// Pop the matching split from the map
+		split := splits[old.Key]
+		delete(splits, old.Key)
+
+		// If the watermark hasn't advanced, there's nothing to do
+		if split.Watermark == nil {
+			continue
+		}
+		if old.Watermark != nil && !old.Watermark.Before(*split.Watermark) {
+			continue
+		}
+
+		// Update the split (the new split's ExecutedOn will be nil, so it will be marked for execution)
+		err = catalog.UpdateModelSplit(ctx, mdl.State.SplitsModelId, split)
+		if err != nil {
+			return fmt.Errorf("failed to update existing split: %w", err)
+		}
+	}
+
+	// The remaining splits are new and should be inserted
+	for _, split := range splits {
+		err = catalog.InsertModelSplit(ctx, mdl.State.SplitsModelId, split)
```

Comment:
Not sure I understand this? Not completely sure, but I don't think a transaction would lead to batching of the inserts (I think it would actually add overhead).

The syncing of splits is designed to be idempotent, so a transaction is not needed from a correctness point of view. If the reconciler is cancelled halfway through the sync, next time it runs it will skip the ones it already added and only sync the ones that are missing (this is also the logic that enables use of splits with `incremental: true`).

Reply:
May be something to be evaluated. But I think if we keep a transaction and then do insert it should give better performance than individual inserts. 
But yeah as noted in other comments lets not worry about the performance here.